src/adapter/src/session.rs

```diff
@@ -102,28 +102,43 @@ impl<T: CoordTimestamp> Session<T> {
         wall_time: DateTime<Utc>,
         access: Option<TransactionAccessMode>,
         isolation_level: Option<TransactionIsolationLevel>,
-    ) -> Self {
+    ) -> (Self, Result<(), AdapterError>) {
+        if let Some(txn) = self.transaction.inner() {
+            if matches!(txn.ops, TransactionOps::Peeks(_))
```

Comment:
What about a tail transaction? That seems like it should also have this behavior? A little concerned that we have to decide on which of these to check or not. Might be safer as a match explicitly handling all states.

Reply:
Ah, that's right/a good call. Fixed locally and will include this in the next push.